examples/pi-home-router/src/arp/arp_handler.rs

```diff
@@ -0,0 +1,212 @@
+    ///         If the pair <protocol type, sender protocol address> is already in my translation
+    ///             table, update the sender hardware address field of the entry with the new
+    ///             information in the packet and set Merge_flag to true.
+    ///         ?Am I the target protocol address?
+    ///         Yes:
+    ///             If Merge_flag is false, add the triplet <protocol type, sender protocol address,
+    ///                 sender hardware address> to the translation table.
+    ///             ?Is the opcode ares_op$REQUEST?  (NOW look at the opcode!!)
+    ///             Yes:
+    ///                 Swap hardware and protocol fields, putting the local hardware and protocol
+    ///                     addresses in the sender fields.
+    ///                 Set the ar$op field to ares_op$REPLY
+    ///                 Send the packet to the (new) target hardware address on the same hardware on
+    ///                     which the request was received.
+    ///
+    fn process(&mut self, packet: Self::Input) -> Option<Self::Output> {
+        let mut arp_frame = ArpFrame::try_from(packet.packet).unwrap();
+        if arp_frame.hardware_type() == ArpHardwareType::Ethernet as u16 {
+            let protocol_type = arp_frame.protocol_type();
+
+            if protocol_type == IPV4_ETHER_TYPE {
+                let mut updated_translation_table = false;
+                let sender_ipv4_address = arp_frame.sender_ipv4_addr().unwrap();
+                let sender_mac_address = arp_frame.sender_mac_addr().unwrap();
+
+                if !self
+                    .arp_table
+                    .lock()
+                    .unwrap()
+                    .contains_key(&sender_ipv4_address)
+                {
+                    self.arp_table
+                        .lock()
+                        .unwrap()
+                        .insert(sender_ipv4_address, sender_mac_address);
+                    updated_translation_table = true;
+                }
+
+                let target_ipv4_address = arp_frame.target_ipv4_addr().unwrap();
+                if target_ipv4_address == ROUTER_IPV4_ADDR {
+                    if !updated_translation_table {
+                        self.arp_table
+                            .lock()
+                            .unwrap()
+                            .insert(sender_ipv4_address, sender_mac_address);
+                    }
+
+                    if arp_frame.opcode() == ArpOp::Request as u16 {
+                        // Generate response ARP frame
+                        arp_frame.set_target_hardware_addr(sender_mac_address);
+                        arp_frame.set_target_protocol_addr(IpAddr::V4(sender_ipv4_address));
+                        // TODO: how can I find this machine's MacAddr?
+                        // Will it be pre-populated in the ARP table?
+                        arp_frame.set_sender_hardware_addr(MacAddr::new([0, 0, 0, 0, 0, 0]));
```

Comment:
Assume that this machine's MAC and IPv4 addresses will be passed in on initialization.

Reply:
You should take the required values as arguments in your new function.  Make sure that the router won't build without the required init variables.  This info will have to be passed into the router, probably via config file, during startup.  I wager rust has a good .toml parser or something lol.